packages/react/src/use-chat.ui.test.tsx

```diff
@@ -556,3 +563,303 @@ describe('maxToolRoundtrips', () => {
     });
   });
 });
+
+describe('file attachments with data url', () => {
+  const TestComponent = () => {
+    const { messages, handleSubmit, handleInputChange, isLoading, input } =
+      useChat({
+        api: '/api/stream-chat',
+      });
+
+    const [attachments, setAttachments] = useState<FileList | undefined>(
+      undefined,
+    );
+    const fileInputRef = useRef<HTMLInputElement>(null);
+
+    return (
+      <div>
+        {messages.map((m, idx) => (
+          <div data-testid={`message-${idx}`} key={m.id}>
+            {m.role === 'user' ? 'User: ' : 'AI: '}
+            {m.content}
+            {m.experimental_attachments?.map(attachment => {
+              if (attachment.contentType?.startsWith('image/')) {
+                return (
+                  <img
+                    key={attachment.name}
+                    className="w-24 rounded-md"
+                    src={attachment.url}
+                    alt={attachment.name}
+                    data-testid={`attachment-${idx}`}
+                  />
+                );
+              } else if (attachment.contentType?.startsWith('text/')) {
+                return (
+                  <div
+                    key={attachment.name}
+                    className="w-32 h-24 rounded-md text-xs ellipsis overflow-hidden p-2 text-zinc-500 border"
+                    data-testid={`attachment-${idx}`}
+                  >
+                    {Buffer.from(
+                      attachment.url.split(',')[1],
+                      'base64',
+                    ).toString('utf-8')}
```

Comment:
this is repeated several times. can we extract it into a helper function that we expose (and maybe test)?

Reply:
done! https://github.com/vercel/ai/pull/2226#discussion_r1678314642